### DISCUSSION TITLE
Drag and drop the newly inserted data does not trigger a business class insert

### DISCUSSION THREAD

**Discussion by angtianqiang** [MarimerLLC/csla#3332] [Questions]
hello Rockford Lhotka:
 I have a question, I hope you help me to answer it
 I have been use CSLA DEV components development, in the past years running is very good, the DEV form has a function can adjust the order, drag and drop the data
 I found: 
drag and drop a data of the original data, save the data triggered when CSLA business class BusinessBase delete and insert
 if insert new drag and drop the data, will not trigger the CSLA business class BusinessBase any CRUD 
why this happen?
I made a DEMO, repeat this kind of question, for BusinessBase CRUD with a breakpoint way to monitor
[WpfApp1.zip](https://github.com/MarimerLLC/csla/files/11417430/WpfApp1.zip)

**Comment by rockfordlhotka** [comment]
I don't know what a "CSLA DEV" component means. There are no real UI components that are part of CSLA.

The CSLA base types don't usually trigger persistence (data portal calls) automatically. The only base type that does this is `DynamicListBase`.

If you have a list of objects in a `BusinessListBase`, you must call the `SaveAsync` method on the list object yourself.

So I am going to assume you are using `DynamicListBase`?

`DynamicListBase` does persist the objects it contains when the user completes editing a row (usually in some sort of data grid UI control). The way this works is that `DynamicListBase` relies on the standard data binding behaviors defined by Microsoft in their Windows Forms data grid control.

Specifically, when the user starts editing a row, the data grid must call `BeginEdit` on the object bound to that row. When the user presses ESC to cancel edits, the data grid must call `CancelEdit` on that object. If the user leaves the row other than pressing ESC, the data grid must call `EndEdit` on that object.

I assume that the items in your list include an "index" property that is used to keep the items in the correct order in the list. So when you drag-and-drop items to rearrange the order of the list, that the "index" properties of all objects in the list are adjusted to reflect the new ordering?

That would mean triggering a save operation on the entire list, or on all the objects affected by the change to those index values.

**Reply by angtianqiang** [reply]
DevExpress DEV here refers to a third party controls, I reconstructed the mentioned in the DEMO code, can help to see when you are free

**Reply by angtianqiang** [reply]
The key word here is drag and drop

**Reply by rockfordlhotka** [reply]
I have a general rule that I don't help support third party UI controls for free - that's an endless time sink that I just can't accept.

I'm explaining the data binding interfaces and behaviors that CSLA expects from a UI control. I've never used this control or its drag-and-drop behavior, but it sounds like it is not automatically calling the necessary begin/cancel/end edit methods on the data binding interface(s).

That does happen, and in such a case hopefully the UI control exposes events that allow _you_ to call those methods to trigger the correct behavior.
